### PR TITLE
Fixes #138773, port autocomplete to OverlayPortal

### LINKED ISSUE #138773: Autocomplete and RawAutocomplete OptionsViewOpenDirection.up not working
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Go to https://api.flutter.dev/flutter/material/Autocomplete-class.html
2. In any of the code samples, add `optionsViewOpenDirection: OptionsViewOpenDirection.up,` to the Autocomplete widget and click Run.
3. Type 'a' in the text field
4. No options show up
5. Change `OptionsViewOpenDirection.up` to `OptionsViewOpenDirection.down`, click Run again and type 'a' in the text field. It works.
6. Basically, the upwards display of the options overlay is not working.
7. Same issue observed with RawAutocomplete too: https://api.flutter.dev/flutter/widgets/RawAutocomplete-class.html
8 Same issue observed locally on Android Studio too.

### Expected results

The options overlay should display upwards when `OptionsViewOpenDirection.up` is set.

### Actual results

Options overlay does not show up

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

/// Flutter code sample for [Autocomplete].

void main() => runApp(const AutocompleteExampleApp());

class AutocompleteExampleApp extends StatelessWidget {
  const AutocompleteExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Autocomplete Basic'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(
                  'Type below to autocomplete the following possible results: ${AutocompleteBasicExample._kOptions}.'),
              const AutocompleteBasicExample(),
            ],
          ),
        ),
      ),
    );
  }
}

class AutocompleteBasicExample extends StatelessWidget {
  const AutocompleteBasicExample({super.key});

  static const List<String> _kOptions = <String>[
    'aardvark',
    'bobcat',
    'chameleon',
  ];

  @override
  Widget build(BuildContext context) {
    return Autocomplete<String>(
      optionsBuilder: (TextEditingValue textEditingValue) {
        if (textEditingValue.text == '') {
          return const Iterable<String>.empty();
        }
        return _kOptions.where((String option) {
          return option.contains(textEditingValue.text.toLowerCase());
        });
      },
      onSelected: (String selection) {
        debugPrint('You just selected $selection');
      },
      optionsViewOpenDirection: OptionsViewOpenDirection.up,
    );
  }
}
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.0, on macOS 13.1 22C65 darwin-arm64, locale en-US)
    • Flutter version 3.16.0 on channel stable at [Removed Directory Path]
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db7ef5bf9f (5 days ago), 2023-11-15 11:25:44 -0800
    • Engine revision 74d16627b9
    • Dart version 3.2.0
    • DevTools version 2.28.2

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at [Removed Directory Path]
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.14.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 119.0.6045.159

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes #138773, port autocomplete to OverlayPortal

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/autocomplete.dart [hunk 1]
```diff
@@ -378,216 +387,122 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
...
-  }
-
-  void _updateActions() {
-    _setActionsEnabled(_focusNode.hasFocus && _selection == null && _options.isNotEmpty);
-  }
-
-  bool _floatingOptionsUpdateScheduled = false;
-  // Hide or show the options overlay, if needed.
-  void _updateOverlay() {
-    if (SchedulerBinding.instance.schedulerPhase == SchedulerPhase.persistentCallbacks) {
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/autocomplete.dart:437]: And great that we no longer have to deal with SchedulerBinding stuff :star_struck:

**Review by justinmc — APPROVED** [review]
LGTM :+1:. It's really cool to see this ported to OverlayPortal, exciting stuff.

As an aside, do you think that the use of OverlayPortal will make it easier to limit the width of the options to the width of the field (https://github.com/flutter/flutter/issues/101620)?

File: packages/flutter/lib/src/widgets/autocomplete.dart [hunk 2]
```diff
@@ -303,68 +302,78 @@ class RawAutocomplete<T extends Object> extends StatefulWidget {
...
-  late final Map<Type, Action<Intent>> _actionMap;
-  late final _AutocompleteCallbackAction<AutocompletePreviousOptionIntent> _previousOptionAction;
-  late final _AutocompleteCallbackAction<AutocompleteNextOptionIntent> _nextOptionAction;
-  late final _AutocompleteCallbackAction<DismissIntent> _hideOptionsAction;
+  final OverlayPortalController _optionsViewController = OverlayPortalController(debugLabel: '_RawAutocompleteState');
+
+  TextEditingController? _internalTextEditingController;
+  TextEditingController get _textEditingController {
+    return widget.textEditingController
+       ?? (_internalTextEditingController ??= TextEditingController()..addListener(_onChangedField));
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/autocomplete.dart:310]: Here?

https://github.com/flutter/flutter/pull/140285/files#diff-4b6f96f28cfdafbb04b797dab70576f40d7866a6291f2981bc7832b475f22355R480-R483

That's only for internal controller/focus nodes. We can't dispose controllers or focus nodes that are not owned by this state.

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/widgets/autocomplete.dart:310]: Yes, I was wondering if you need to call `_internalTextEditingController?.removeListener(_onChangedField);` there in addition to disposing it.

[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/autocomplete.dart:310]: Oh sorry I misread that. The dispose method gets rid of all listeners.

[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/widgets/autocomplete.dart:310]: Ok got it, thanks for clarifying :+1:

**Pull request merged**